core/governance/market.go

```diff
@@ -231,6 +235,27 @@ func validateFuture(future *types.FutureProduct, decimals uint64, assets Assets,
 		return types.ProposalErrorInvalidFutureProduct, ErrMissingOracleSpecForTradingTermination
 	}
 
+	for i, f := range future.OracleSpecForTradingTermination.ToOracleSpec().Filters {
+		if f.Key.Type == oraclespb.PropertyKey_TYPE_TIMESTAMP {
```

Comment:
I think it makes sense to do this validation when there is a timestamp condition in oracle to catch it as soon as possible, but not all oracles will have a timestamp property. And those could still terminate a market before enactment while circumventing this check.

So I imagine we'll still need some code somewhere to catch that scenario and work out what to do when terminating that orcale submission comes in when the market isn't enacted (ignore it, apply it and stop, something else)

Reply:
You mean that not all oracles will have a "PropertyKey_TYPE_TIMESTAMP" set type for the `key` field, but still have a timestamp for the `value` field
AND
will contain a filter for `OracleSpecForTradingTermination`?

Reply:
As far as I see now the property name "vegaprotocol.builtin.timestamp" is used to set also its type `oraclespb.PropertyKey_TYPE_TIMESTAMP` and the callback that uses both values inside the filter key for making the subscription here:
[[OracleSpecForTradingTermination](https://github.com/vegaprotocol/vega/blob/develop/core/products/future.go#L266)](https://github.com/vegaprotocol/vega/blob/develop/core/products/future.go#L266)

Maybe you mean that in `toTimestampCondiiton` - we receive only a condition

Reply:
The `vegaprotocol.builtin.timestamp` oracles will be fine and caught here in this new check. So thats good.

Non-time-triggered oracles will just come from a transacstion submission from whatever public-keys were in the spec. For example here are the oracle-specs on testnet for terminated/settling a market: https://explorer.fairground.wtf/oracles and the property values are all other `TypeBoolean` or `TypeInteger`.

In this case its then possible for someone to propose a market with enactment in like 5 days time, but for the oracle submissions for termination/settlement to be submitted after 4 days. Then when we hit day 5 it panics because we try to enact a market in a unexpected state. This is what (I'm assuming) happened here: https://vegaprotocol.slack.com/archives/CEE07A084/p1658934399014629

The market was settled, cleaned up and removed from the execution engine and then when enactment came we panicked because the market didn't exist anymore: 
```
processor/abci.go:1282        unable to start market opening auction        {"market-id": "89021b6e5c6bff671918ea93acd256fef79984303d6b5bcf40cf13cdbc1a8824", "error": "market does not exist"}
```

Reply:
Agree that early trading termination cannot be prevented with validation at market creation time. It’s definitely the case that some data sources will have no timestamp at all or the market may for some reason want to filter on a date/time value that’s in the past.

That means that it is absolutely possible to receive a valid (by the market and data source definition) settlement or trading terminated trigger event before the market is in a state that can receive it, and it needs to be processed safely and reasonably.

The [market lifecycle spec](https://github.com/vegaprotocol/specs/blob/master/protocol/0043-MKTL-market_lifecycle.md) details all the valid stage transitions. It should not be possible to exit the proposed or pending states except in the conditions and to the states defined in the market lifecycle. 

For example, this spec states that after the enactment date, if trading terminated is triggered, the market becomes [Cancelled](https://github.com/vegaprotocol/specs/blob/master/protocol/0043-MKTL-market_lifecycle.md#cancelled). (Note that settlement cannot occur unless the market is in trading terminated as per [here](https://github.com/vegaprotocol/specs/blob/master/protocol/0016-PFUT-product_builtin_future.md#42-final-settlement-expiry) and [here](https://github.com/vegaprotocol/specs/blob/master/protocol/0043-MKTL-market_lifecycle.md#settled).)

It’s possible we need to add some clarity to the spec on when to process and when to reject/ignore data sources in certain states. It’s also possible that we’ve missed some edge cases or complexities and should revisit the current design…